andela/travel_tool_back
# #160047854 User Roles API

#### What does this PR do?
 - add a ends to sent user roles by the super admin and add new users to the database
- A  /user endpoint to add a new user POST
- A /user/:id to view one user GET
- A /user:id to change the role of the user PUT
- A /user to view all users in the database GET
- A /role to view all roles in the database GET
- A /role to add a new role POST
- A /role/:id to view a single role GET

#### Description of Task to be completed?
- A new user can be added to the database
- A new role can be added to the database by the super admin
- The super admin can assign a role to a user or update the ufsr role
- An endpoint to get the list of users and roles in the database

#### How should this be manually tested?
- Clone the repo -` git clone https://github.com/andela/travel_tool_back.git`
- Enter into the project directory - `cd travel_tool_back`
- Start the app using `make start`
- Open `Postman`  
- To add new user send a `POST` request to the endpoint 
`http://local.andela.com:5000/api/v1/user` 
  with email and full name field 
- To get a list of all users send a `GET` request to the endpoint 
`http://local.andela.com:5000/api/v1/user` 
  to see a list of all users in the database
- To get one user, send a `GET` request to the endpoint with the user id
`http://local.andela.com:5000/api/v1/user/:id` 
  to see a single user in the database
- To get add a new role for a user, send a `PUT` request to the endpoint with the user id
`http://local.andela.com:5000/api/v1/user/:id`  and a field with roleId and the role id
  to update the user role
- To add new role send a `POST` request to the endpoint 
`http://local.andela.com:5000/api/v1/role` 
  with roleName and description field 
- To get a list of all roles send a `GET` request to the endpoint 
`http://local.andela.com:5000/api/v1/role` 
  to see a list of all roles in the database
- To get one role, send a `GET` request to the endpoint with the role id
`http://local.andela.com:5000/api/v1/role/:id` 
  to see a single role with user associations in the database
- To set the default super admin add the email to the donenv file, send a `GET` request to the 
 endpoint `http://local.andela.com:5000/api/v1/admin` 

#### Any background context you want to provide?
- Only the Super admin can change user role and add new roles to the database. The default role when a new user sign up is `Requester`
- All the endpoint requires authentication. To include token in the `authorization` header:
- Login to the [travela app](https://travela-staging.andela.com)
- After you are logged in, visit the developer console in your browser and open `Application`, you should see your `jwt-token` there.
- Copy the token and add it to the `authorization` header when making a request in postman.
- The endpoint to change user role and add new role to the dadabase requres the user to be a super admin
- The set default super admin endpoint will only make the person a super admin if the user email address in the dotenv file and the one from the jwt token are the same and the user already exist in the database

#### What are the relevant pivotal tracker stories?
[[#160047854]](https://www.pivotaltracker.com/n/projects/2184887/stories/160047854)

#### Screenshots (if appropriate)
![image](https://user-images.githubusercontent.com/20597418/44779054-5b3cc500-ab76-11e8-8a1e-67cf33581f1e.png)
![image](https://user-images.githubusercontent.com/20597418/44779061-5f68e280-ab76-11e8-82b1-c7cb4f2cc7ae.png)
![image](https://user-images.githubusercontent.com/20597418/44779071-62fc6980-ab76-11e8-8248-cbc443960af7.png)


#### Questions:
